crates/ruff_server/src/session/workspace/configuration.rs

```diff
@@ -0,0 +1,74 @@
+use anyhow::anyhow;
+use lsp_types::Url;
+use ruff_workspace::resolver::{ConfigurationTransformer, Relativity};
+use std::{
+    collections::BTreeMap,
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+#[derive(Default)]
+pub(crate) struct RuffConfiguration {
+    // settings to pass into the ruff linter
+    pub(crate) linter: ruff_linter::settings::LinterSettings,
+    // settings to pass into the ruff formatter
+    pub(crate) formatter: ruff_workspace::FormatterSettings,
+}
+
+#[derive(Default)]
+pub(super) struct ConfigurationIndex {
+    index: BTreeMap<PathBuf, Arc<RuffConfiguration>>,
+}
+
+impl ConfigurationIndex {
+    pub(super) fn get_or_insert(&mut self, document_url: &Url) -> Arc<RuffConfiguration> {
+        let document_path = document_url
+            .to_file_path()
+            .expect("document URL should be a valid path");
+        let folder = document_path
+            .parent()
+            .expect("document URL should be a file path and have a parent");
+        if let Some(config) = self.index.get(folder) {
+            return config.clone();
+        }
+
+        let config = Arc::new(Self::find_configuration_at_path(folder).unwrap_or_else(|err| {
+            tracing::error!("The following error occurred when trying to find a configuration file at `{}`:\n{err}", document_path.display());
+            tracing::error!("Falling back to default configuration for `{}`", document_path.display());
+            RuffConfiguration::default()
+        }));
+
+        self.index.insert(folder.to_path_buf(), config.clone());
+
+        config
+    }
+
+    pub(super) fn clear(&mut self) {
+        self.index.clear();
+    }
+
+    fn find_configuration_at_path(folder: &Path) -> crate::Result<RuffConfiguration> {
+        let pyproject = ruff_workspace::pyproject::find_settings_toml(folder)?
+            .ok_or_else(|| anyhow!("No pyproject.toml/ruff.toml/.ruff.toml file was found"))?;
+        let settings = ruff_workspace::resolver::resolve_root_settings(
```

Comment:
Calling `find_settings_toml` is probably fine for the LSP but it means that it traverses the entire ancestory chain for each directory once, and parse the settings. That means for projects with a single configuration file that we'll reparse (and store in meory) the same configuration once for every possible directory. 

I think I would prefer if we either:

* Discover all setting files during startup and setup the configuration index. The index only contains entries for directories that contain a configuration. Resolving a configuration is the same as before. Use the `BTreeMap` to find the closest directory that contains a configuration. I think that's my preferred solution because it also "just works" when deleting a configuration. All that is necessary is to remove (or add) the deleted (or added) configuration from the `ConfigurationsIndex` and it isn't necessary to invalidat any sub paths. It may also avoid the need for using an `Arc`.
* Or we change the logic here to iterate over the anchestor in the `ConfigurationIndex` starting at the current folder and see if it contains any configuration. If yes, parse it and set it in the Index. If not, go up the ancestory chain and check the parent directory (after testing that the index doesn't already contains an entry for this path).